sdk/src/transaction.rs

```diff
@@ -246,6 +255,26 @@ impl Transaction {
             .collect())
     }
 
+    /// Replace all the signatures and pubkeys
+    pub fn replace_signatures(&mut self, signers: &[(Pubkey, Signature)]) -> Result<()> {
+        let num_required_signatures = self.message.header.num_required_signatures as usize;
+        if signers.len() != num_required_signatures
+            || self.signatures.len() != num_required_signatures
+            || self.message.account_keys.len() < num_required_signatures
+        {
+            return Err(TransactionError::InvalidAccountIndex);
+        }
+
+        signers
+            .iter()
+            .enumerate()
+            .for_each(|(i, (pubkey, signature))| {
+                self.signatures[i] = *signature;
+                self.message.account_keys[i] = *pubkey;
```

Comment:
I don't think account keys should be overridden. They should be checked for equality. Otherwise the compiled instructions could point to the wrong account key.

Reply:
Where do the account keys come from if the online entity doesn't have the keypairs (only the offline entity does)?  Are you thinking the pubkeys should come from a config file?

Couple of thoughts
- account keys are public, if this is a legitimate concern we should probably make those private
- The account keys used in the instructions themselves don't come from the config file (private keypairs), they come from parameters passed on the command line (there may be exceptions).  
- If the account keys change then the signatures would have to still match and if they do then its what the signer intended anyway.

Reply:
Hmm good points.

Could account keys come from just the public keys then and when you pass in signers, you don't attempt to sign the transaction?

> account keys are public, if this is a legitimate concern we should probably make those private

Yeah, I think I agree

> The account keys used in the instructions themselves don't come from the config file (private keypairs), they come from parameters passed on the command line (there may be exceptions).

I'm confused by this. Looks like you are still creating a signed transaction with config here: https://github.com/solana-labs/solana/pull/7104/files#diff-6897771877027eaa520b55b24b6dceafR797. That doesn't look right... Can you help me understand why this makes sense?

> If the account keys change then the signatures would have to still match and if they do then its what the signer intended anyway.

Yeah true, but could lead to confusing errors because it wouldn't be obvious that the signature was failing because of mis-ordered signers right?

Reply:
We don't really have a way to determine whether signer ordering is correct since we don't have anything to compare to between what the CLI caller passes in and what is in the transaction.

As for why we create a signed transaction, the instructions themselves don't contain the pubkey, just the indices (unless they are placed into the instruction data but that's a whole different issue) and therefore get replaced appropriately.  If we are worried about non-signer keys we could replace all the account keys.  A signed transaction matches the source form and order that the pubkey=signature pairs were provided to the offline user when the original was produced.  I would rather not generate the transaction differently whether its offline or online.  If for some use case we encounter a case where the private key is stored in the instruction data we can deal with that appropriately then.

Reply:
Ah ok! Makes a lot more sense now, thanks for explaining.